### PR TITLE
repository: (Mostly) index-related cleanups

### PR DESCRIPTION
**Pull request by MichaelEischer**
What does this PR change? What problem does it solve?
-----------------------------------------------------
This PR simplifies some of the index code and removes several functions that are only used for testing. In addition, it slightly improves the separation between the index and repository code. These changes are nearly sufficient to allow splitting the index from the repository package.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------
No.

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added tests for all code changes.
- ~~[ ] I have added documentation for relevant changes (in the manual).~~
- ~~[ ] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).~~
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

### CODE REVIEW

File: internal/repository/index.go [hunk 1]
```diff
@@ -223,22 +181,6 @@ func (idx *Index) Has(bh restic.BlobHandle) bool {
 	return idx.byType[bh.Type].get(bh.ID) != nil
 }
 
-// LookupSize returns the length of the plaintext content of the blob with the
```
Comment (greatroar) [inline internal/repository/index.go:226]: Are you sure we don't want to keep a separate path for looking up sizes? It short-circuits when there are duplicates and it doesn't allocate.

**Comment by greatroar** [comment]
repository.Loader is also unused.

File: internal/restic/repository.go [hunk 2]
```diff
@@ -4,6 +4,7 @@ import (
 	"context"
 
 	"github.com/restic/restic/internal/crypto"
+	"github.com/restic/restic/internal/ui/progress"
```
Comment (greatroar) [inline internal/restic/repository.go:7]: I think this is a sign that (after this PR) we should drop this file and the [single-implementation interfaces](https://medium.com/@cep21/preemptive-interface-anti-pattern-in-go-54c18ac0668a) in it, and just let client code work on *repository.Repository instead. Changes to that type already require changes here, and now every package that wants to handle a restic.ID drags in ui/progress. That's not a very big deal yet, but it does mean that ui/progress cannot be told about restic.ID when the need arises.

Comment (MichaelEischer) [reply internal/restic/repository.go:7]: We currently have a few tests that wrap `restic.Repository` for the checker and the archiver. These wouldn't be possible anymore if we drop the interface. Although the wrapper repos only use `LoadTree`, `LoadBlob` and `LookupBlobSize`. Replacing these wrappers with different interfaces would probably be rather messy.

Dragging in the progress.Counter is a bit unfortunate, but if necessary we can decouple that with an additional interface ;-) .

I also plan to split the restic package into a backend part, the snapshot data model part and the repository related parts. There's a bunch of totally experimental commits at https://github.com/MichaelEischer/restic/tree/layered-repo2 which includes the just mentioned refactoring and a rather random collection of other stuff.

File: internal/repository/index.go (hunk 1 above)
Comment (MichaelEischer) [reply internal/repository/index.go:226]: I'm not too happy that `LookupSize` has to reimplement `Blob.DataLength()`, but I'm not settled on getting rid of that function or not.

File: internal/restic/repository.go (hunk 2 above)
Comment (greatroar) [reply internal/restic/repository.go:7]: If we define checker.Repository and archiver.Repository with the set of operations those packages need, then IMO that's not messy at all. It signals that checker needs one set of operations and archiver another. There's a motto "[accept interfaces, return structs](https://nathanleclaire.com/blog/2015/10/10/interfaces-and-composition-for-effective-unit-testing-in-golang/)", and although it's [controversial](https://github.com/go-proverbs/go-proverbs.github.io/issues/37), I think it applies here.

This is all beside the point, which is that your PR LGTM. Testing the waters for a further refactor, which I'll postpone. Looking forward to your follow-up :)

Comment (fd0) [reply internal/restic/repository.go:7]: Cleaning up all my old beginner mistakes, awesome! :)

File: internal/repository/index.go (hunk 1 above)
Comment (MichaelEischer) [reply internal/repository/index.go]: I've removed the commit for now.

**Comment by MichaelEischer** [comment]
So, then let's get this merged.

**Pull request merged**